### PR TITLE
ci: Fix codecov uploads

### PR DESCRIPTION
**Pull request by zharinov**
## Changes

- Split the coverage task to the `codecov` and `coverage-threshold` parts
- Revert `codecov-action` to v3, which allows to upload reports created from fork PRs

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by zharinov** [comment]
It looks like we're reverting to v3, until the issues are fixed on the Codecov side

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -373,35 +373,48 @@ jobs:
...
-
       - name: Download coverage reports
         uses: actions/download-artifact@9bc31d5ccc31df68ecc42ccf4149144866c47d8a # v3.0.2
         with:
           name: coverage
           path: coverage
 
       - name: Codecov
-        uses: codecov/codecov-action@e0b68c6749509c5f83f984dd99a76a1c1a231044 # v4.0.1
+        uses: codecov/codecov-action@ab904c41d6ece82784817410c45d8b8c02684457 # v3.1.6
```
Comment (viceice) [inline .github/workflows/build.yml:389]: 🤔 why it should work for PR's from forks without token. Do you find a issue somewhere about the broken behavior?

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -373,35 +373,48 @@ jobs:
...
+  coverage-threshold:
+    needs: [test]
+    runs-on: ubuntu-latest
+    timeout-minutes: 3
+    if: (success() || failure()) && github.event.pull_request.draft != true
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11 # v4.1.1
+        with:
+          sparse-checkout: ${{ env.SPARSE_CHECKOUT }}
```
Comment (viceice) [inline .github/workflows/build.yml:404]:
```suggestion
          sparse-checkout: ${{ env.SPARSE_CHECKOUT }}
          filter: blob:none # we don't need all blobs
```

reduce git download size

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -373,35 +373,48 @@ jobs:
...
           path: coverage
 
       - name: Codecov
-        uses: codecov/codecov-action@e0b68c6749509c5f83f984dd99a76a1c1a231044 # v4.0.1
+        uses: codecov/codecov-action@ab904c41d6ece82784817410c45d8b8c02684457 # v3.1.6
         with:
           directory: coverage/lcov
-          token: ${{ secrets.CODECOV_TOKEN }}
+          fail_ci_if_error: true
+          verbose: true
```
Comment (viceice) [inline .github/workflows/build.yml:393]:
```suggestion
```

Isn't this too much output?

**Review by viceice — COMMENTED** [review]
will be conflicted after #27268 merge

File: .github/workflows/build.yml (hunk 3 above)
Comment (zharinov) [reply .github/workflows/build.yml:393]: Why not? If we don't need it, we don't care. If we need it, it's better to have all of it.

File: .github/workflows/build.yml (hunk 1 above)
Comment (zharinov) [reply .github/workflows/build.yml:389]: - https://github.com/codecov/feedback/issues/112

Comment (zharinov) [reply .github/workflows/build.yml:390]: Maybe, it'll help to add the secret to Dependabot secrets, even we're not using it

Comment (zharinov) [reply .github/workflows/build.yml:390]: But anyway, at this point I'd prefer to fix it first, then upgrade and check

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -419,7 +434,7 @@ jobs:
 
   # Catch-all required check for test matrix and coverage
   test-success:
-    needs: [setup, test, coverage]
+    needs: [setup, test, codecov, coverage-threshold]
```
Comment (viceice) [inline .github/workflows/build.yml:437]:
```suggestion
    needs: 
      - setup
      - test
      - codecov
      - coverage-threshold
```

we should use array style for less future changes

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -392,17 +387,37 @@ jobs:
...
-        uses: codecov/codecov-action@e0b68c6749509c5f83f984dd99a76a1c1a231044 # v4.0.1
+        uses: codecov/codecov-action@ab904c41d6ece82784817410c45d8b8c02684457 # v3.1.6
         with:
           directory: coverage/lcov
-          token: ${{ secrets.CODECOV_TOKEN }}
+          fail_ci_if_error: true
+          verbose: true
+
+  coverage-threshold:
+    needs: [test]
```
Comment (viceice) [inline .github/workflows/build.yml:397]:
```suggestion
    needs:
      - test
```

**Review by viceice — DISMISSED** [review]

**Review by viceice — APPROVED** [review]

**Pull request merged**